### ISSUE TITLE
Unhandled Runtime Error Error: ERR_ALREADY_CALLED

### ISSUE THREAD

**Issue by kcbuilder** [Telegram-Mini-Apps/nextjs-template#19]
Failed to launch application using latest version of template

## Steps to Reproduce 
* created telegram bot 
* created application using template, https://github.com/kcbuilder/tg-app-example - public version is here
* started local version with `pnpm run dev`
* started local tunnel with ngrok `ngrok http 3000`
* set menu link for created bot  in BotFather 

When trying to open bot in telegram getting problem 
![image](https://github.com/user-attachments/assets/fc4999f2-2f70-4253-a6eb-e38e3a524c45)

Tested for application created with same template a few month ago, all working good 
![image](https://github.com/user-attachments/assets/a82a86ad-e157-4350-9a01-576d6933e35c)

**Comment by heyqbnk** [comment]
I wasn't able to reproduce the problem using the nextjs-template, nor the repository you provided. It just opens as usual for me :(

**Comment by braveltd** [comment]
Fixed by

```
import {
  backButton,
  viewport,
  themeParams,
  miniApp,
  initData,
  $debug,
  init as initSDK,
} from "@telegram-apps/sdk-react";

/**
 * Initializes the application and configures its dependencies.
 */
export function init(debug: boolean): void {
  // Set @telegram-apps/sdk-react debug mode.
  $debug.set(debug);

  // Initialize special event handlers for Telegram Desktop, Android, iOS, etc.
  // Also, configure the package.
  initSDK();

  // Mount all components used in the project.
  if (backButton.isSupported()) {
    backButton.mount();
  }

  // Define components-related CSS variables.
  if (!miniApp.isMounted()) {
    miniApp.mount();
    miniApp.bindCssVars();
  }
  if (!themeParams.isMounted()) {
    themeParams.mount();
    themeParams.bindCssVars();
  }

  initData.restore();

  if (!viewport.isMounted() && !viewport.isMounting()) {
    void viewport.mount().catch((e) => {
      console.error("Something went wrong mounting the viewport", e);
    });
  }

  if (viewport.isMounted()) {
    viewport.bindCssVars();
  }
}
```

**Comment by kcbuilder** [comment]
> Fixed by changing core/init.ts to
> 
> ```
> import {
>   backButton,
>   viewport,
>   themeParams,
>   miniApp,
>   initData,
>   $debug,
>   init as initSDK,
> } from "@telegram-apps/sdk-react";
> 
> /**
>  * Initializes the application and configures its dependencies.
>  */
> export function init(debug: boolean): void {
>   // Set @telegram-apps/sdk-react debug mode.
>   $debug.set(debug);
> 
>   // Initialize special event handlers for Telegram Desktop, Android, iOS, etc.
>   // Also, configure the package.
>   initSDK();
> 
>   // Mount all components used in the project.
>   if (backButton.isSupported()) {
>     backButton.mount();
>   }
> 
>   // Define components-related CSS variables.
>   if (!miniApp.isMounted()) {
>     miniApp.mount();
>     miniApp.bindCssVars();
>   }
>   if (!themeParams.isMounted()) {
>     themeParams.mount();
>     themeParams.bindCssVars();
>   }
> 
>   initData.restore();
> 
>   if (!viewport.isMounted() && !viewport.isMounting()) {
>     void viewport.mount().catch((e) => {
>       console.error("Something went wrong mounting the viewport", e);
>     });
>   }
> 
>   if (viewport.isMounted()) {
>     viewport.bindCssVars();
>   }
> }
> ```
> 
> In my project's i don't use eruda, but you can include eruda, i don't think it will break anythink!

Got problem with applying theme for dark mode with this approach, not sure about correlation here but would like to share this case
![image](https://github.com/user-attachments/assets/47bc5218-e31f-41bd-bd8c-2c39db5b4349)

**Comment by braveltd** [comment]
> > Fixed by changing core/init.ts to
> > ```
> > import {
> >   backButton,
> >   viewport,
> >   themeParams,
> >   miniApp,
> >   initData,
> >   $debug,
> >   init as initSDK,
> > } from "@telegram-apps/sdk-react";
> > 
> > /**
> >  * Initializes the application and configures its dependencies.
> >  */
> > export function init(debug: boolean): void {
> >   // Set @telegram-apps/sdk-react debug mode.
> >   $debug.set(debug);
> > 
> >   // Initialize special event handlers for Telegram Desktop, Android, iOS, etc.
> >   // Also, configure the package.
> >   initSDK();
> > 
> >   // Mount all components used in the project.
> >   if (backButton.isSupported()) {
> >     backButton.mount();
> >   }
> > 
> >   // Define components-related CSS variables.
> >   if (!miniApp.isMounted()) {
> >     miniApp.mount();
> >     miniApp.bindCssVars();
> >   }
> >   if (!themeParams.isMounted()) {
> >     themeParams.mount();
> >     themeParams.bindCssVars();
> >   }
> > 
> >   initData.restore();
> > 
> >   if (!viewport.isMounted() && !viewport.isMounting()) {
> >     void viewport.mount().catch((e) => {
> >       console.error("Something went wrong mounting the viewport", e);
> >     });
> >   }
> > 
> >   if (viewport.isMounted()) {
> >     viewport.bindCssVars();
> >   }
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > In my project's i don't use eruda, but you can include eruda, i don't think it will break anythink!
> 
> Got problem with applying theme for dark mode with this approach, not sure about correlation here but would like to share this case ![image](https://private-user-images.githubusercontent.com/134141967/379350717-47bc5218-e31f-41bd-bd8c-2c39db5b4349.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk3MDQxNTQsIm5iZiI6MTcyOTcwMzg1NCwicGF0aCI6Ii8xMzQxNDE5NjcvMzc5MzUwNzE3LTQ3YmM1MjE4LWUzMWYtNDFiZC1iZDhjLTJjMzlkYjViNDM0OS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDIzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAyM1QxNzE3MzRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jNjU1ZWZmZTJiN2JjYzU4MDMyMTAyNWE2MDhhODc5YzE5NjU5MmE3ZGFkYmUwNzYxZWYwZGYzMmRlMjVlZjRjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.DOewDUWMXN93YnslhQKzk4lUsG_3SC8nVvYmOXhIVkA)

probably i should move all bindCss after mount's. checking if mounted. like with viewport!

**Comment by braveltd** [comment]
In my current project, i have only dark mode possible, so i don't know if it's my init.ts code that breaks or somethink else. for me, what i provided works with nextjs@14.2.16.

**Comment by heyqbnk** [comment]
The problem is probably related to the React's StrictMode performing all effects twice. It is probably resetting refs also, we use them in the `useTelegramMock` (`useClientOnce` actually) function mocking the environment.

To solve it, we need to call the `init` function only once before the React tree is rendered, or at least unmount all mounted SDK components.

I currently have no time for it and can fix the problem only on the next week. Feel free to propose a solution

**Comment by vladutjs** [comment]
Has anybody got around this issue?

**Comment by miguelaeh** [comment]
I can confirm that this is the fix: https://github.com/Telegram-Mini-Apps/nextjs-template/issues/19#issuecomment-2429284188

**Comment by vikiival** [comment]
Based on I managed to get it working [#19 (comment)](https://github.com/Telegram-Mini-Apps/nextjs-template/issues/19#issuecomment-2429284188)

The diff is that I placed `themeParams.bindCssVars();` outside of the if statement 🤷 

additionally I have added 
`import '@telegram-apps/telegram-ui/dist/styles.css';` in `Root.tsx`

### TL;DR 

```ts
export function init(debug: boolean): void {
  // Set @telegram-apps/sdk-react debug mode.
  $debug.set(debug);

  // Initialize special event handlers for Telegram Desktop, Android, iOS, etc.
  // Also, configure the package.
  initSDK();

  // Mount all components used in the project.
  if (backButton.isSupported()) {
    backButton.mount();
  }

  // Define components-related CSS variables.
  if (!miniApp.isMounted()) {
    miniApp.mount();
    miniApp.bindCssVars();
  }
  if (!themeParams.isMounted()) {
    themeParams.mount();
  }
  
  themeParams.bindCssVars();

  initData.restore();

  if (!viewport.isMounted() && !viewport.isMounting()) {
    void viewport.mount().catch((e) => {
      console.error("Something went wrong mounting the viewport", e);
    });
  }

  if (viewport.isMounted()) {
    viewport.bindCssVars();
  }
}
```

**Comment by braveltd** [comment]
> Based on I managed to get it working [#19 (comment)](https://github.com/Telegram-Mini-Apps/nextjs-template/issues/19#issuecomment-2429284188)
> 
> The diff is that I placed `themeParams.bindCssVars();` outside of the if statement 🤷
> 
> additionally I have added `import '@telegram-apps/telegram-ui/dist/styles.css';` in `Root.tsx`
> ### TL;DR
> 
> ```ts
> export function init(debug: boolean): void {
>   // Set @telegram-apps/sdk-react debug mode.
>   $debug.set(debug);
> 
>   // Initialize special event handlers for Telegram Desktop, Android, iOS, etc.
>   // Also, configure the package.
>   initSDK();
> 
>   // Mount all components used in the project.
>   if (backButton.isSupported()) {
>     backButton.mount();
>   }
> 
>   // Define components-related CSS variables.
>   if (!miniApp.isMounted()) {
>     miniApp.mount();
>     miniApp.bindCssVars();
>   }
>   if (!themeParams.isMounted()) {
>     themeParams.mount();
>   }
>   
>   themeParams.bindCssVars();
> 
>   initData.restore();
> 
>   if (!viewport.isMounted() && !viewport.isMounting()) {
>     void viewport.mount().catch((e) => {
>       console.error("Something went wrong mounting the viewport", e);
>     });
>   }
> 
>   if (viewport.isMounted()) {
>     viewport.bindCssVars();
>   }
> }
> ```

i believe you can make PR with that code. at least this works for me. importing styles helped a lot. or we should just wait for react-ui, react-sdk to get fixed!

**Comment by braveltd** [comment]
Made an PR for that issue...
